### ISSUE TITLE
Constrained domain adaptation for labeling of intervertebral discs

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/ivadomed#550]
The paper by [Bateson et al.](https://arxiv.org/pdf/1908.02996.pdf) seems relevant for this problem.

**Comment by rezazad68** [comment]
I would like to follow the below steps to accomplish this task: please correct me if I am not in the correct direction.

1- Run the Bateson et al [method](https://github.com/CDAMICCAI2019/CDA) on the same dataset as reported in the paper [IVDM3Seg](https://ivdm3seg.weebly.com/data.html).
2- Revise the code for Spine Generic public database

There is an issue for the first step, the dataset access requires confirmation which I already sent but I did not get a response yet. It would be great in case you have access to data I can use, as a part of this project. 
I am working on the Spine Generic public database to completely understand the modalities and details through the documentation. I would update in case if I face issues. 


Your feedback would be very helpful, thanks.

**Comment by jcohenadad** [comment]
@rezazad68 we do have the iVDM3Seg dataset in the lab, but i think you can directly jump to step number 2

**Comment by rezazad68** [comment]
Following the [documentation](https://spine-generic.readthedocs.io/en/latest/analysis-pipeline.html) I have processed the dataset, the resulted "spineGeneric_results" folder contains several modalities and information for each person. The method proposed by [Bateson et al](https://github.com/CDAMICCAI2019/CDA) uses the Water (Wat) modality as the Source domain, and the In-Phase (IP) modality as the Target domain. For [spine-generic dataset](https://github.com/spine-generic/data-multi-subject#spine-generic-public-database-multi-subject) which modalities should be used as domain and target? could you please determine the related modalities and segmentation files?

**Comment by jcohenadad** [comment]
labeling should be done on the T1w and T2w. @lrouhier can give you more context

**Comment by rezazad68** [comment]
I have reviewed most of the samples for T1w and T2w in "SpineGeneric_results" but did not find the 3D segmentation masks. There is only a 2D disc mask is available, however, for the training process, 3D segmentation masks are required. I also reviewed the `index.html` file in quality control, which provides a user-friendly visualization tool but did not find 3D masks for discs labeling. @lrouhier could you please advise me? 

The method proposed by [Bateson et al](https://github.com/CDAMICCAI2019/CDA) uses 2D slices, thus, I need a 3D disc mask to generate such slices for the deep model.

**Comment by jcohenadad** [comment]
Indeed, there are no 3D segmentation mask as the task is semantic labeling (ie: single voxel per disc). Please see the work of @lrouhier (and Lucas: please add some relevant material here).

**Comment by lrouhier** [comment]
Hi, sorry about the delay. 
There are no 3D segmentation masks as Julien mention. As single voxel labels represent a really small target (even in 2D), we took inspiration from the detection of facial keypoint (see [Wang et al](https://arxiv.org/pdf/1904.07399.pdf)) And converted the single voxel to Gaussian functions. I think this could work in 3D as well, to generate the masks.
The [Forsberg et al](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5537089/) article is interesting as it is taking a different approach, even though it might not be the most optimized algorithm. Basically, you would need a dataset with segmented discs, but we could use a model to segment the disc and then retrieve what we want from there. This would allow you to use the same method as Bentson. 
Another option would be to use a cascaded architecture. The first step would be an R-CNN to get a large sense of the localisation of each disc (see [Sa et al](https://ieeexplore.ieee.org/abstract/document/8036887?casa_token=wYq24QMDw7cAAAAA:zmNBGuadcao1DnRYR6Uh6VTiZ3gyy_YIhB3rdSlaZK2h2ttg1r3zny41quzowQiT5bW0SbL2SqU)). This would be the constrained step. Then we could extract these patches and feed them to a small countception/Attention Unet/CNN architecture to get the point at the posterior tip of the disc.


About the ground truth: 
You can use the spine-generic/data-multi-subject. I created my 2D ground truth for the loader with the script called [prepare_dataset_vertebral_labeling](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/scripts/prepare_dataset_vertebral_labeling.py). This was tailored to my needs But I am sure we can modify it to make it 3D or more relvant for this.

**Comment by jcohenadad** [comment]
> You can use the spine-generic/data-multi-subject. I created my 2D ground truth for the loader with the script called prepare_dataset_vertebral_labeling. This was tailored to my needs But I am sure we can modify it to make it 3D or more relvant for this.

I would suggest to use the same ground truth, so the method @rezazad68 will implement could be directly compared (that's the whole point-- otherwise comparison not possible). @rezazad68 the idea here is to adapt the method so it is compatible with single-voxel labels

**Comment by rezazad68** [comment]
@jcohenadad Thanks for the update, I would follow your direction. To get an idea of how they use the dataset I have prepared IVDM3Seg. dataset according to the requirements, now the problem is the missing code (dice3dn) in the [Bateson et al](https://github.com/CDAMICCAI2019/CDA) GitHub. I have created an issue and contacted the author still waiting for an update. I will follow the same pre-processing approach on Spine Generic public database and will revise it for semantic labeling.
